### ISSUE TITLE
Generated headers are not order-only deps for generated source files

### ISSUE THREAD

**Issue by kjellahl** [mesonbuild/meson#6391]
The description of executable() in the reference manual, https://mesonbuild.com/Reference-manual.html, says

>With the Ninja backend, Meson will create a build-time order-only dependency on all generated input files

I suppose this is also how library() is supposed to work. I can't get this to work as
expected with library() when both the source file (.cc file) and the header files are generated.

For .cc files in the source tree (not generated) it works as expected. The generated
headers become order-only dependencies in the build.ninja file (listed after '||').
For .cc files generated with custom_target(), all generated header files, listed as
sources in library(), become implicit dependencies (listed after '|').

We have projects with many generated .cc files and many generated header files.
The way Meson handles those files, all the generated .cc files are recompiled when
a generated header file has been changed, whether they include the changed header
file or not.

I use meson 0.52.1 and ninja 1.9.0 on Ubuntu 19.10.

**Comment by kjellahl** [comment]
A patch such as this one fixes the issue for me.
```
diff --git a/mesonbuild/backend/ninjabackend.py b/mesonbuild/backend/ninjabackend.py
index 824b9584..d81f870a 100644
--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -578,8 +578,7 @@ int dummy;
             if self.environment.is_llvm_ir(src):
                 o = self.generate_llvm_ir_compile(target, src)
             else:
-                o = self.generate_single_compile(target, src, True,
-                                                 header_deps=header_deps)
+                o = self.generate_single_compile(target, src, True, [], header_deps)
             obj_list.append(o)
 
         use_pch = self.environment.coredata.base_options.get('b_pch', False)
```
Can it have bad side effects in other cases? Was the generated headers deliberately
added as implicit dependencies to the generated sources? It has been like that since
commit ccf63459c6c9090518ce9164c67bc12de010bc21 2014-08-24.

**Comment by kjellahl** [comment]
If the present behavior is deliberate, is there a workaround?
It causes many unnecessary recompilations. It's obviously contrary to what's
said in Meson's documentation, that it's meant to be extremely fast.

**Comment by marc-h38** [comment]
> all the generated .cc files are recompiled when a generated header file has been changed, whether they include the changed header file or not.

This seems to be the real issue: why are these generated .cc files depending _at all_ on headers files not included and irrelevant?

Yes "Order-only" saves recompilations, but only because it's an ugly optimization hack. "Order-only" means: "I need this dependency but I promise updates to it don't matter. Any version will do. So create it if it's missing but ignore its timestamp for the sake of speed" Now cross fingers that future updates will actually not matter, otherwise the crash might really hurt!

But if some headers are not included at all, why are they dependencies at all?!? There's something worse than an order-only issue in your case.

**Comment by kjellahl** [comment]
In many C++ modules in Gnome, most .cc and .h files are generated. Where Meson is
used, they are generated with custom_target(). They are then used for building a
library file. Example:
```
pangomm_library = library(pangomm_pcname,
  built_cc_file_targets, built_h_file_targets, extra_cc_files,
  version: pangomm_libversion,
  include_directories: extra_include_dirs,
  dependencies: pangomm_build_dep,
  install: true,
)
```
built_cc_file_targets and built_h_file_targets are arrays with about 25 elements each.
extra_cc_files is an array with names of non-generated .cc files.

From the Meson Reference Manual, https://mesonbuild.com/Reference-manual.html,
description of executable():

>With the Ninja backend, Meson will create a build-time order-only dependency on all
>generated input files, including unknown files. For all input files (generated and
>non-generated), Meson uses the dependency file generated by your compiler to
>determine when to rebuild sources. The behavior is similar for other backends.

I assume that this should hold also for library().

What actually happens is that the .cc files in extra_cc_files get an order-only
dependency on all generated header files, as expected. But the generated .cc files
in built_cc_file_targets get an implicit dependency on all generated header files.
Why are generated and non-generated .cc files treated differently in this respect?

>"Order-only" means: "I need this dependency but I promise updates to it don't matter.
>Any version will do. So create it if it's missing but ignore its timestamp for
>the sake of speed"

This is not how it's described in https://ninja-build.org/manual.html#ref_dependencies:

>3. _Order-only dependencies_, expressed with the syntax _|| dep1 dep2_ on the end of a
>build line. When these are out of date, the output is not rebuilt until they are built,
>but changes in order-only dependencies alone do not cause the output to be rebuilt. 

>why are these generated .cc files depending at all on headers files not included and irrelevant?

Because I don't want to manually make lists of header files that each .cc file
depends on. The compiler can do that, and does it more reliably than I do.
That's the _dependency file generated by your compiler_ that Meson and Ninja rely on.

**Comment by marc-h38** [comment]
> Because I don't want to manually make lists of header files that each .cc file
depends on.

I wasn't asking who defines the dependencies. I'm asking whether there are spurious dependencies or not.

> the compiler can do that, and does it more reliably than I do.

Doesn't seem very reliable if there are `foo.cc` -> `bar.h` dependencies where `foo.c` doesn't include `bar.h`. Seems spurious whether it's order-only or not.

**Comment by kjellahl** [comment]
No, no, no. The .cc files do not depend on header files that they don't include
(directly or indirectly via other header files).

When a library is built, many generated .cc files are compiled. Each .cc file
depends on a subset of the generated .h files, different subsets for different .cc files.
It is error-prone to manually keep track of which .cc files depend on which header files.
And it's unnecessary, because the compiler can do that.

If it can be guaranteed that all generated header files that need updating,
are updated before any .cc file is compiled, the compilations are always made with
updated header files. I think this can be achieved with order-only dependencies
without necessarily causing all generated .cc files to be re-compiled. You say order-only
dependency only guarantees that the generated header files exist when the .cc files
are compiled, but that it does not guarantee that the header files are updated.

I still don't understand why generated and non-generated .cc files are treated
differently. With the library() function call in my previous comment, generated .cc
files get an implicit dependency on the generated .h files, non-generated .cc files
get an order-only dependency.

**Comment by marc-h38** [comment]
> The .cc files do not depend on header files that they don't include (directly or indirectly via other header files).

Yet at the very end of the description you wrote:

> all the generated .cc files are recompiled when a generated header file has been changed, **whether they include the changed header file or not.**

Why are the .cc files that do NOT depend **AT ALL** on the changed header recompiled? I'm sorry but that's a much more serious issue than any order-only issue. Do these .cc files show **ANY** sort of spurious dependency in build.ninja?

PS: no need to rephrase the order-only issue yet another time, my question is NOT about order-only.

**Comment by kjellahl** [comment]
But _my_ question is about order-only dependency, or rather the lack of it.

When both generated .cc files and generated .h files are listed as sources in
a library() function call, meson tells ninja that all the listed generated .cc files
depend implicitly on all the listed generated .h files. That is why ninja re-compiles
all the .cc files when any one of the .h files is updated.

An excerpt from a build.ninja file shows the difference between generated and
non-generated .cc files.
tabarray.cc and all the shown .h files are generated with custom_target().
tabarray.cc gets implicit dependencies (|).
init.cc is read from the source tree. It gets order-only dependencies (||).
```
build pango/pangomm/c9b4a7d@@pangomm-2.44@sha/meson-generated_.._tabarray.cc.o: cpp_COMPILER pango/pangomm/tabarray.cc | pango/pangomm/attributes.h pango/pangomm/attriter.h pango/pangomm/attrlist.h pango/pangomm/cairofontmap.h pango/pangomm/color.h pango/pangomm/context.h pango/pangomm/coverage.h pango/pangomm/font.h pango/pangomm/fontdescription.h pango/pangomm/fontface.h pango/pangomm/fontfamily.h pango/pangomm/fontmap.h pango/pangomm/fontmetrics.h pango/pangomm/fontset.h pango/pangomm/glyph.h pango/pangomm/glyphstring.h pango/pangomm/item.h pango/pangomm/language.h pango/pangomm/layout.h pango/pangomm/layoutiter.h pango/pangomm/layoutline.h pango/pangomm/layoutrun.h pango/pangomm/rectangle.h pango/pangomm/renderer.h pango/pangomm/tabarray.h
 DEPFILE = pango/pangomm/c9b4a7d@@pangomm-2.44@sha/meson-generated_.._tabarray.cc.o.d
 ARGS = -Ipango/pangomm/c9b4a7d@@pangomm-2.44@sha -Ipango/pangomm -I../../gnome/pangomm/pango/pangomm -Ipango/pangomm/.. -I../../gnome/pangomm/pango/pangomm/.. -I/opt/gnome/include/glibmm-2.64 -I/opt/gnome/lib/glibmm-2.64/include -I/opt/gnome/include/glib-2.0 -I/opt/gnome/lib/glib-2.0/include -I/opt/gnome/include/sigc++-3.0 -I/opt/gnome/lib/sigc++-3.0/include -I/opt/gnome/include/cairomm-1.16 -I/opt/gnome/lib/cairomm-1.16/include -I/opt/gnome/include/cairo -I/opt/gnome/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/opt/gnome/include/pango-1.0 -I/opt/gnome/include/fribidi -I/opt/gnome/include/harfbuzz -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O2 -g -pedantic -Wall -Wextra -Wformat-security -Wsuggest-override -Wshadow -Wno-long-long -Werror -DG_DISABLE_DEPRECATED -DPANGO_DISABLE_DEPRECATED -DGLIBMM_DISABLE_DEPRECATED -DSIGCXX_DISABLE_DEPRECATED -fPIC

build pango/pangomm/c9b4a7d@@pangomm-2.44@sha/init.cc.o: cpp_COMPILER ../../gnome/pangomm/pango/pangomm/init.cc || pango/pangomm/attributes.h pango/pangomm/attriter.h pango/pangomm/attrlist.h pango/pangomm/cairofontmap.h pango/pangomm/color.h pango/pangomm/context.h pango/pangomm/coverage.h pango/pangomm/font.h pango/pangomm/fontdescription.h pango/pangomm/fontface.h pango/pangomm/fontfamily.h pango/pangomm/fontmap.h pango/pangomm/fontmetrics.h pango/pangomm/fontset.h pango/pangomm/glyph.h pango/pangomm/glyphstring.h pango/pangomm/item.h pango/pangomm/language.h pango/pangomm/layout.h pango/pangomm/layoutiter.h pango/pangomm/layoutline.h pango/pangomm/layoutrun.h pango/pangomm/rectangle.h pango/pangomm/renderer.h pango/pangomm/tabarray.h
 DEPFILE = pango/pangomm/c9b4a7d@@pangomm-2.44@sha/init.cc.o.d
 ARGS = -Ipango/pangomm/c9b4a7d@@pangomm-2.44@sha -Ipango/pangomm -I../../gnome/pangomm/pango/pangomm -Ipango/pangomm/.. -I../../gnome/pangomm/pango/pangomm/.. -I/opt/gnome/include/glibmm-2.64 -I/opt/gnome/lib/glibmm-2.64/include -I/opt/gnome/include/glib-2.0 -I/opt/gnome/lib/glib-2.0/include -I/opt/gnome/include/sigc++-3.0 -I/opt/gnome/lib/sigc++-3.0/include -I/opt/gnome/include/cairomm-1.16 -I/opt/gnome/lib/cairomm-1.16/include -I/opt/gnome/include/cairo -I/opt/gnome/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/opt/gnome/include/pango-1.0 -I/opt/gnome/include/fribidi -I/opt/gnome/include/harfbuzz -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O2 -g -pedantic -Wall -Wextra -Wformat-security -Wsuggest-override -Wshadow -Wno-long-long -Werror -DG_DISABLE_DEPRECATED -DPANGO_DISABLE_DEPRECATED -DGLIBMM_DISABLE_DEPRECATED -DSIGCXX_DISABLE_DEPRECATED -fPIC
```

**Comment by marc-h38** [comment]
> But my question is about order-only dependency,

**Comment by marc-h38** [comment]
> But my question is about order-only dependency,

https://en.wikipedia.org/wiki/XY_problem

"order-only" is meant to optimize valid dependencies. It's not meant to mitigate bogus dependencies.

Thanks for sharing an actual example, it helps. Does `init.cc` really include all these headers? Directly or not. Are these headers not already in init.cc's DEPFILE? (`ninja -d keepdepfile`) Can you share the meson code for init.cc?

**Comment by marc-h38** [comment]
Answering my own question at last and correcting myself after some more research (and thanks again for sharing the build.ninja)

> It's not meant to mitigate bogus dependencies.

Sometimes it can be. "order-only" is also designed to "bootstrap" a depfile.

The depfile is generated by the compiler. To generate a depfile, the compiler needs headers to at least exist. The list of headers that needs to exist (for each .cc) can be found in ...the depfile that's not generated yet! Chicken-and-egg situation. Brute force solution: make each .cc depend on all generated headers. Then mitigate these massive and bogus dependencies by making them all order-only. Finally, append the depfile with the real dependencies to patch up the whole mess.

Sorry this took so long, order-only documentation is incredibly limited. Hopefully this will help others.

**Comment by marc-h38** [comment]
Now back to your issue

> I still don't understand why generated and non-generated .cc files are treated differently.

Assuming the *generated .cc* and the *generated .h* are generated by the same `custom_target()`, there's no depfile bootstrap issue there. By the time the compiler is invoked to generate the true dependencies in the depfile, the generated .h files are already there. So IMHO there should be a "different treatment" but not the current one: instead there should be no dependency **at all** in `build.ninja` between *generated .cc* and the *generated .h*. The depfile is enough as usual.

**Comment by kjellahl** [comment]
>Does init.cc really include all these headers?

No. Neither does tabarray.cc.

>Can you share the meson code for init.cc?

The project is stored at https://gitlab.gnome.org/GNOME/pangomm/
The relevant meson.build file is pango/pangomm/meson.build at
https://gitlab.gnome.org/GNOME/pangomm/blob/master/pango/pangomm/meson.build
In this file you can ignore the part that starts with `else # not maintainer_mode`.
When maintainer_mode is false, there are no generated source files, neither .cc files
nor .h files.

Two requirements on `execute()`, `library()` and similar functions:
- No source file shall be compiled before all included header files have been created or (if needed) updated.
- No source file shall be recompiled unless either the source file itself or an included file has been updated.

You say that order-only dependency is a really bad way of achieving these goals.
Can you suggest a better way?

Let's study an unrealistically simple example:
Generated files: g1.cc, g1.h, g2.cc, g2.h, g3.cc, g3.h
File read from the source tree: s1.cc
There may also be header files included from the source tree and from elsewhere,
but they are of no interest here.

g1.cc includes g1.h and g2.h
g2.cc includes g2.h
g3.cc includes g3.h
s1.cc includes g2.h
No header files include other header files.

Each .cc file depends on the .h files that it includes.

Assume that after some updates of the files from which .cc and .h files are
generated, g2.cc, g2.h and g3.h have been changed, and now g2.h includes g3.h.
After these updates g1.cc and s1.cc depend on g3.h, which they did not before.
g1.cc and s1.cc must not be compiled before g3.h has been updated.

The first requirement is achieved if all compilation is postponed until all
generated .h files have been created or (if necessary) updated.
Ninja's or other backend's freedom to choose the order of executing commands is
then restricted more than is absolutely necessary. IMO that's acceptable.

**Comment by marc-h38** [comment]
> You say that order-only dependency is a really bad way of achieving these goals.
> Can you suggest a better way?

Yes, the one _you_ gave: the depfile. Order-only deps are only to boostrap the depfile.

> After these updates g1.cc and s1.cc depend on g3.h, which they did not before.

**Comment by kjellahl** [comment]
The depfile shows a .cc file's dependencies when it was last compiled, doesn't it?
The list of dependencies may change, e.g. as in the example in my previous comment.
If ninja would rely only on depfiles during the hypothetical update run, the
following sequence of events would be possible:

Update g2.h
Compile g1.cc: Oops, now g1.cc depends on g3.h. Check if it needs updating.
Update g3.h
Compile g1.cc again

Is ninja really that smart? It's even possible that the first compilation of g1.cc
fails, because the updated g2.h is not compatible with the old version of g3.h.

The depfile is excellent for deciding which files must be recompiled, but is it enough
for deciding which generated headers must be updated before the compiler is invoked?

I've a feeling that this discussion does not lead us closer to a solution to the
original problem: Too many files are recompiled when a generated header file is updated.

**Comment by kjellahl** [comment]
I think I've found a workaround:
```
  built_dummy_h_file_target = custom_target('dummy_header.h',
    input: built_h_file_targets,
    output: 'dummy_header.h',
    command: [
      python3, project_source_root / 'tools' / 'dummy-header.py',
      '@OUTPUT0@',
    ],
    build_by_default: maintainer_mode,
    install: false,
  )

  extra_include_dirs = ['..']
  pangomm_library = library(pangomm_pcname,
    built_cc_file_targets, extra_cc_files, built_dummy_h_file_target,
    version: pangomm_libversion,
    include_directories: extra_include_dirs,
    dependencies: pangomm_build_dep,
    install: true,
  )
```
If dummy_header.h does not exist, it's created, but it's never updated.

dummy_header.h pretends to depend on all the generated header files. It's the only
generated header file in the library() function call. It doesn't matter what kind
of dependency the .cc files have on dummy_header.h as long as there is some kind
of dependency. All generated headers will be updated before anything is compiled.
Only necessary recompilations are made, thanks to depfile.

The workaround is fairly simple. I can live with it. But workarounds are always ugly.
I would prefer if some smart fix in meson can make it unnecessary.

**Comment by marc-h38** [comment]
> The depfile shows a .cc file's dependencies when it was last compiled, doesn't it?

Did you just try it? `ninja -v -d keepdefile ...`

> Oops, now g1.cc depends on g3.h. Check if it needs updating. ... Is ninja really that smart? It's even possible that the first compilation of g1.cc fails,...

It's not rocket science for ninja to notice the depfile it parses every time has changed. In that case I suspect it might rebuild everything, after all the graph of includes and dependencies is not supposed to change frequently, Again if you have an actual example it should be very easy to try.

> I've a feeling that this discussion does not lead us closer to a solution to the
original problem: Too many files are recompiled when a generated header file is updated.

Above I recommended to get rid of of all meson-added dependencies between files generated by the same `custom_target`

> I would prefer if some smart fix in meson can make it unnecessary.

The reason depfiles exist is because only compilers have  knowledge of  includes. Meson cannot (and shouldn't) be smart about includes because it is one layer away from it.

**Comment by kjellahl** [comment]
It's clear that Meson won't be changed the way I proposed. Closing this issue.

**Comment by kjellahl** [comment]
A fix equivalent to the one that I suggested was added with commit https://github.com/mesonbuild/meson/commit/a25f0741e8ac663c8b3a6e33df8a0875dceab804.
It's included in meson 0.55.0.
